### PR TITLE
Np 595 update template

### PR DESCRIPTION
**Pull request by svenngjeto**

### CODE REVIEW

File: buildspec.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+version: 0.2
+
+phases:
+
+  install:
+    runtime-versions:
+      java: openjdk8
```
Comment (janbj) [inline buildspec.yml]: openjdk11?

File: buildspec_publish_to_sar.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+version: 0.2
+
+phases:
+  install:
+    runtime-versions:
+      java: openjdk8
```
Comment (janbj) [inline buildspec_publish_to_sar.yml]: openjdk11?

File: src/main/java/no/unit/nva/alma/AlmaRecordParser.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            TransformerException, SAXException, ParserConfigurationException, XPathExpressionException {
+
+        Reference reference = new Reference();
+        try (InputStream inputStream = new ByteArrayInputStream(
+                CharStreams.toString(inputStreamReader).getBytes(StandardCharsets.UTF_8))) {
+
+            Record record = getFirstMarcRecord(inputStream);
+
+            if (isNull(record)) {
+                return reference;
```
Comment (janbj) [inline src/main/java/no/unit/nva/alma/AlmaRecordParser.java]: Why do we want to return a Reference with no title?

File: src/main/java/no/unit/nva/alma/AlmaRecordParser.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            }
+
+            DataField datafield = (DataField) record.getVariableField(MARC_DATAFIELD_245);
+            String title = datafield.getSubfields().stream()
```
Comment (janbj) [inline src/main/java/no/unit/nva/alma/AlmaRecordParser.java]: Extract to method and give it a more descriptive name?

File: src/main/java/no/unit/nva/alma/AlmaRecordParser.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                    .collect(Collectors.toMap(Subfield::getCode, Subfield::getData))
+                    .entrySet().stream().sorted(Map.Entry.comparingByKey())
+                    .map(Map.Entry::getValue).collect(Collectors.joining(BLANK));
+
+            reference.setTitle(title);
+            return reference;
+        }
+    }
+
+    private boolean filterSubfields(Subfield subfield) {
```
Comment (janbj) [inline src/main/java/no/unit/nva/alma/AlmaRecordParser.java]: Rename method to subfields we are filtering for?

File: src/main/java/no/unit/nva/alma/AlmaRecordParser.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        documentBuilderFactory.setNamespaceAware(true);
+        DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
+        Document document = documentBuilder.parse(inputStream);
+        document.getDocumentElement().normalize();
+        XPath xpath = XPathFactory.newInstance().newXPath();
+        xpath.setNamespaceContext(new NamespaceResolver(document));
+        Node element = (Node) xpath.evaluate(MARC_RECORD_XPATH, document.getDocumentElement(), XPathConstants.NODE);
+
+        if (isNull(element)) {
+            return null;
```
Comment (janbj) [inline src/main/java/no/unit/nva/alma/AlmaRecordParser.java]: return Optional?

File: src/main/java/no/unit/nva/alma/AlmaRecordParser.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+            XPathExpressionException, IOException, SAXException, ParserConfigurationException {
+
+        DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+        documentBuilderFactory.setNamespaceAware(true);
+        DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
+        Document document = documentBuilder.parse(inputStream);
+        document.getDocumentElement().normalize();
+        XPath xpath = XPathFactory.newInstance().newXPath();
+        xpath.setNamespaceContext(new NamespaceResolver(document));
+        Node element = (Node) xpath.evaluate(MARC_RECORD_XPATH, document.getDocumentElement(), XPathConstants.NODE);
```
Comment (janbj) [inline src/main/java/no/unit/nva/alma/AlmaRecordParser.java]: Extract some of this to new method?

File: src/main/java/no/unit/nva/alma/AlmaRecordParser.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+            return null;
+        }
+
+        Document result = documentBuilder.newDocument();
+        Node collection = result.createElementNS(MARC_NAMESPACE, COLLECTION_ELEMENT);
+        result.appendChild(collection);
+        Node node = result.importNode(element, true);
+        result.getElementsByTagNameNS(MARC_NAMESPACE, COLLECTION_ELEMENT).item(FIRST_NODE).appendChild(node);
+
+        Source source = new DOMSource(result);
```
Comment (janbj) [inline src/main/java/no/unit/nva/alma/AlmaRecordParser.java]: Extract some of this to new method that processes result?

File: src/main/java/no/unit/nva/alma/AlmaSruConnection.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    public static final String RECORD_SCHEMA_KEY = "recordSchema";
+    public static final String RECORD_SCHEMA = "marcxml";
+    public static final String MAXIMUM_RECORDS_KEY = "maximumRecords";
+    public static final String MAX_NUMBER_RECODS = "2";
+    public static final String START_RECORD_KEY = "startRecord";
+    public static final String START_RECORD_1 = "1";
+    public static final String QUERY_KEY = "query";
+
+    protected InputStreamReader connect(URL url) throws IOException {
+        return new InputStreamReader(url.openStream());
```
Comment (janbj) [inline src/main/java/no/unit/nva/alma/AlmaSruConnection.java:29]: Can we use new Java http client?

File: src/main/java/no/unit/nva/alma/FetchAlmaRecordHandler.java [hunk 10]
```diff
@@ -40,9 +60,24 @@ public GatewayResponse handleRequest(final Map<String, Object> input, Context co
...
+            try (InputStreamReader streamReader = connection.connect(queryUrl)) {
+                AlmaRecordParser almaRecordParser = new AlmaRecordParser();
+                Reference json = almaRecordParser.extractPublicationTitle(streamReader);
+                gatewayResponse.setBody(gson.toJson(json, Reference.class));
+                gatewayResponse.setStatusCode(Response.Status.OK.getStatusCode());
+            }
+        } catch (URISyntaxException | IOException | TransformerException | SAXException | ParserConfigurationException
+                | XPathExpressionException e) {
+            System.out.println(e);
+            gatewayResponse.setErrorBody(e.getMessage());
```
Comment (janbj) [inline src/main/java/no/unit/nva/alma/FetchAlmaRecordHandler.java]: Method that maps exception to response?

File: src/main/java/no/unit/nva/alma/FetchAlmaRecordHandler.java [hunk 11]
```diff
@@ -40,9 +60,24 @@ public GatewayResponse handleRequest(final Map<String, Object> input, Context co
...
-        //String creatorName = queryStringParameters.get(CREATOR_NAME_KEY);
+        Map<String, String> queryStringParameters = (Map<String, String>) input.get(QUERY_STRING_PARAMETERS_KEY);
+        String scn = queryStringParameters.get(SCN_KEY);
+        String creatorName = queryStringParameters.get(CREATOR_NAME_KEY);
+        Gson gson = new GsonBuilder().setPrettyPrinting().create();
+        try {
+            final URL queryUrl = connection.generateQueryUrl(scn, creatorName);
+            try (InputStreamReader streamReader = connection.connect(queryUrl)) {
+                AlmaRecordParser almaRecordParser = new AlmaRecordParser();
+                Reference json = almaRecordParser.extractPublicationTitle(streamReader);
```
Comment (janbj) [inline src/main/java/no/unit/nva/alma/FetchAlmaRecordHandler.java:73]: Can we extract this logic to a service?

File: src/test/java/no/unit/nva/alma/AlmaRecordParserTest.java [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    public void testExtractPublicationTitle_FromZeroHits() throws IOException, ParserConfigurationException,
+            TransformerException, SAXException, XPathExpressionException {
+        InputStream stream = AlmaRecordParserTest.class.getResourceAsStream(SRU_RESPONSE_ZERO_HITS);
+        InputStreamReader inputStreamReader = new InputStreamReader(stream);
+        AlmaRecordParser almaRecordParser = new AlmaRecordParser();
+        Reference reference = almaRecordParser.extractPublicationTitle(inputStreamReader);
+        assertNull(reference.getTitle());
+    }
+
+    @Test(expected = SAXException.class)
```
Comment (janbj) [inline src/test/java/no/unit/nva/alma/AlmaRecordParserTest.java]: You can Junit 5 assert throws

**Review by janbj — CHANGES_REQUESTED** [review]

File: buildspec.yml (hunk 1 above)
Comment (svenngjeto) [reply buildspec.yml]: Yess, openjdk11

File: buildspec_publish_to_sar.yml (hunk 2 above)
Comment (svenngjeto) [reply buildspec_publish_to_sar.yml]: ...openjdk11

File: src/main/java/no/unit/nva/alma/AlmaRecordParser.java (hunk 5 above)
Comment (svenngjeto) [reply src/main/java/no/unit/nva/alma/AlmaRecordParser.java]: Done, filterSubfieldsAorB

**Pull request merged**